src/go/k8s/internal/controller/redpanda/redpanda_controller.go

```diff
@@ -531,390 +524,294 @@ func setHelmLabelsAndAnnotations(object client.Object, rp *v1alpha2.Redpanda) {
+			continue
 		}
-		if helmUpgrade.Remediation != nil {
-			upgrade.Remediation = helmUpgrade.Remediation
+		err := r.Delete(ctx, resource)
+		if err != nil {
+			// Do not stop deletion when one resource can not be deleted.
+			// It might be the case where resource was already deleted.
+			log.Error(err, "failed to delete resource", "resource", resource)
 		}
 	}
 
-	return &helmv2beta2.HelmRelease{
-		ObjectMeta: metav1.ObjectMeta{
-			Name:            rp.GetHelmReleaseName(),
-			Namespace:       rp.Namespace,
-			OwnerReferences: []metav1.OwnerReference{rp.OwnerShipRefObj()},
-		},
-		Spec: helmv2beta2.HelmReleaseSpec{
-			Chart: helmv2beta2.HelmChartTemplate{
-				Spec: helmv2beta2.HelmChartTemplateSpec{
-					Chart:    "redpanda",
-					Version:  rp.Spec.ChartRef.ChartVersion,
-					Interval: &metav1.Duration{Duration: 1 * time.Minute},
-					SourceRef: helmv2beta2.CrossNamespaceObjectReference{
-						Kind:      "HelmRepository",
-						Name:      rp.GetHelmRepositoryName(),
-						Namespace: rp.Namespace,
-					},
-				},
-			},
-			Values:   values,
-			Interval: metav1.Duration{Duration: 30 * time.Second},
-			Timeout:  timeout,
-			Upgrade:  upgrade,
-		},
-	}, nil
-}
-
-func (r *RedpandaReconciler) createHelmRepositoryFromTemplate(rp *v1alpha2.Redpanda) *sourcev1.HelmRepository {
-	return &sourcev1.HelmRepository{
-		ObjectMeta: metav1.ObjectMeta{
-			Name:            rp.GetHelmRepositoryName(),
-			Namespace:       rp.Namespace,
-			OwnerReferences: []metav1.OwnerReference{rp.OwnerShipRefObj()},
-		},
-		Spec: sourcev1.HelmRepositorySpec{
-			Interval: metav1.Duration{Duration: 30 * time.Second},
-			URL:      v1alpha2.RedpandaChartRepository,
-		},
+	patch := client.MergeFrom(rp.DeepCopy())
+	controllerutil.RemoveFinalizer(rp, FinalizerKey)
+	if err := r.Patch(ctx, rp, patch); err != nil {
+		log.Error(err, "finalizer not removed")
 	}
-}
 
-func (r *RedpandaReconciler) patchRedpandaStatus(ctx context.Context, rp *v1alpha2.Redpanda) error {
-	key := client.ObjectKeyFromObject(rp)
-	latest := &v1alpha2.Redpanda{}
-	if err := r.Client.Get(ctx, key, latest); err != nil {
-		return err
-	}
-	return r.Client.Status().Patch(ctx, rp, client.MergeFrom(latest))
+	return ctrl.Result{}, nil
 }
 
-// event emits a Kubernetes event and forwards the event to notification controller if configured.
-func (r *RedpandaReconciler) event(rp *v1alpha2.Redpanda, revision, severity, msg string) {
-	var metaData map[string]string
-	if revision != "" {
-		metaData = map[string]string{v2.GroupVersion.Group + revisionPath: revision}
+func (r *RedpandaReconciler) reconcileStatus(ctx context.Context, rp *v1alpha2.Redpanda) error {
+	log := ctrl.LoggerFrom(ctx).WithName("RedpandaReconciler.reconcileStatus")
+
+	rp.Status.ObservedGeneration = rp.Generation
+
+	dot, err := rp.GetDot()
+	if err != nil {
+		log.Error(err, "failed to get dot")
+	}
+	sel := redpanda.StatefulSetPodLabelsSelector(dot)
+
+	// Get redpanda Pods readiness
```

Comment:
If you rebase, can we re-use andrew's new logic here?

Reply:
Yes, I will adopt this status report to use statefulset and deployment. The more important discussion I would like to have is how our Redpanda Status CRD should look like. Currently only condition with ObservedGeneration indicate the reconciliation. 

I would like to settle that it's enough for any potential user (rpk, kubectl, argocd/flux, terraform). From experience the kubernetes terraform provider has somewhat problematic way to "wait for particular status to appear".